Changes CustomHTTPErrors annotation to use custom default backend

jasongwartz opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide/pull-requests.md#the-pull-request-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
-->

**What this PR does / why we need it**:

This PR makes it possible to use a default-backend annotation together with a custom-http-errors annotation. When the default-backend annotation is set, the upstream of the `@custom` error locations is set to the upstream name of the default-backend annotation's service.

Cases covered:
- single ingress object for a hostname, both default-backend and custom-http-errors annotations set
- multiple ingress objects for the same hostname, different paths, each with their own set of default-backend and custom-http-errors annotations (eg. capturing errors to one default-backend for a certain path vs another default-backend, or the global default-backend, for a different path - via separate ingress objects but the same hostname)

This involved making the name of the error-handling location `@custom_<upstream-name>_<code>`. This is needed for the case where, within one server block, different paths (locations) are capturing the same error code but sending it to different default-backends. This means that the new 'default' error location name is `@custom_upstream-default-backend_<code>`.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: 

This PR is a follow-up to https://github.com/kubernetes/ingress-nginx/pull/3344, specifically addressing comment: https://github.com/kubernetes/ingress-nginx/pull/3344#issuecomment-436238468. This also resolves https://github.com/kubernetes/ingress-nginx/issues/3700

**Special notes**:

I've had some trouble running the e2e-tests locally so I might need to revisit those. Thanks for reviewing!

jasongwartz commented on internal/ingress/controller/controller.go:
```diff
@@ -591,11 +591,13 @@ func (n *NGINXController) getBackendServers(ingresses []*ingress.Ingress) ([]*in
 
 								nb := upstream.DeepCopy()
-								name := fmt.Sprintf("custom-default-backend-%v", upstream.Name)
+								name := fmt.Sprintf("custom-default-backend-%v", location.DefaultBackend.GetName())
```
This change means that the backend name of custom-error locations is `custom-default-backend-<service_name>` where `service_name` is the name of the default-backend service. It used to be that the `service_name` would be the name of the upstream for which this default-backend is substituted. Eg. where it used to be `custom-default-backend-my-app`, it is now `custom-default-backend-app-errorhandler`.

This is necessary since multiple locations within a server block can share the same 'default-backend' - so now, the name of the error-handler-location clearly signifies which service it will route to.

ElvinEfendi replied:
Is this a necessary change? I wonder what was the rationale to design that way initially? Could it be maybe because it would explicitly mention what upstream it is replacing due to the fact that upstream has no endpoints?

jasongwartz replied:
well it's necessary if multiple ingress objects (ie locations) for a single upstream can have different error blocks. if I don't make this change, all locations that point to a given upstream have to share the same default-backend.

I assume it was that way initially since the default-backend was only for when the upstream had no endpoints, thus this upstream was replacing the regular upstream entirely (whereas now, if endpoints > 0, the default backend is just another upstream).

ElvinEfendi replied:
Given two ingresses one with location /one pointing to upstream X and another with location /two pointing to upstream X as well. And upstream X has no endpoints. Then without this change we can not use different `default-backend` annotation per location, right?

Given that this change is not necessarily about custom errors, do you think we should introduce it in a different PR with its own tests and documentation adjustments?

jasongwartz replied:
the case I was more targeting is two ingresses /one and /two, same upstream, both with custom error annotations but different default backends - but yeah, that case you mentioned is also true.

jasongwartz replied:
It might be tricky to pull it out into another PR - I'll look and see how tightly I integrated the rest. I don't think it will have any user-facing change, but we can definitely add tests and update the docs (https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#default-backend is out-of-date anyway)

jasongwartz commented on internal/ingress/controller/template/template.go:
```diff
@@ -882,41 +884,87 @@ func proxySetHeader(loc interface{}) string {
-func buildCustomErrorDeps(proxySetHeaders map[string]string, errorCodes []int, enableMetrics bool) interface{} {
+func buildCustomErrorDeps(upstreamName string, errorCodes []int, enableMetrics bool) interface{} {
 	return struct {
-		ProxySetHeaders map[string]string
```
As far as I could see, the ProxySetHeaders map was unused in the template. If this is a bug, we can re-add it and fix.

jasongwartz commented:
/assign bowei

aledbf commented:
@jasongwartz please don't assign PRs manually

jasongwartz commented:
@aledbf sorry, my bad. can we remove that line from the k8s-ci-robot message somehow?

aledbf commented:
> can we remove that line from the k8s-ci-robot message somehow?

No, because the bot reads the owners files and tries to suggest a name :disappointed:

aledbf commented:
@jasongwartz please squash the commits

jasongwartz commented:
@aledbf cool, have done so

aledbf commented:
/approve

aledbf commented:
@ElvinEfendi please review (and lgtm)

ElvinEfendi commented on internal/ingress/controller/controller.go:
```diff
@@ -580,7 +580,7 @@ func (n *NGINXController) getBackendServers(ingresses []*ingress.Ingress) ([]*in
 				if upstream.Name == location.Backend {
-					if len(upstream.Endpoints) == 0 {
+					if len(upstream.Endpoints) == 0 || len(location.CustomHTTPErrors) != 0 {
 						klog.V(3).Infof("Upstream %q has no active Endpoint", upstream.Name)
```
I think this need to be replaced - it is not going to be true when `len(upstream.Endpoints) != 0 and len(location.CustomHTTPErrors) != 0`

jasongwartz replied:
it should be true in that case, right? maybe I'm not following, but if you mean `false || true` it should eval to true

ElvinEfendi commented on internal/ingress/controller/controller.go:
```diff
@@ -591,11 +591,13 @@ func (n *NGINXController) getBackendServers(ingresses []*ingress.Ingress) ([]*in
 									location.Path, server.Hostname, location.DefaultBackend.Namespace, location.DefaultBackend.Name)
```
similar to https://github.com/kubernetes/ingress-nginx/pull/3730/files#r256678420, this log message can be lie, maybe move it to https://github.com/kubernetes/ingress-nginx/pull/3730/files#diff-86ac9ff9d75a0c5005c116e766a6127dR598?

jasongwartz replied:
done as well, good catch - I had missed that entirely

ElvinEfendi commented on internal/ingress/controller/controller.go:
```diff
@@ -580,7 +580,7 @@ func (n *NGINXController) getBackendServers(ingresses []*ingress.Ingress) ([]*in
+					if len(upstream.Endpoints) == 0 || len(location.CustomHTTPErrors) != 0 {
 						klog.V(3).Infof("Upstream %q has no active Endpoint", upstream.Name)
 						// check if the location contains endpoints and a custom default backend
 						if location.DefaultBackend != nil {
```
What do you think of merging everything until https://github.com/kubernetes/ingress-nginx/pull/3730/files#diff-86ac9ff9d75a0c5005c116e766a6127dR590 and extracting them all together into a helper function something called `shouldCreateUpstreamForLocationDefaultBackend`?

jasongwartz replied:
done, let me know if that's what you had in mind

ElvinEfendi commented on internal/ingress/controller/template/template.go:
```diff
@@ -882,41 +884,87 @@ func proxySetHeader(loc interface{}) string {
+	if location.DefaultBackend != nil {
+		return fmt.Sprintf("custom-default-backend-%v", location.DefaultBackend.GetName())
+	}
+	return "upstream-default-backend"
```
I think these values are used in controller.go as well, how hard would it be to put them into a helper function and constant respectively and reuse here?

jasongwartz replied:
I'd be happy to do that. package controller has a utils file already, but importing the controller package here doesn't quite seem right - is there somewhere with shared utils (across both controller and template)?

ElvinEfendi replied:
Maybe we can make this function a property of `Location` struct?

I don't mind leaving this as is for now.

jasongwartz replied:
I tried an implementation of making it a property of Location struct, let me know if it's what you had in mind?

ElvinEfendi commented on internal/ingress/controller/template/template.go:
```diff
@@ -882,41 +884,87 @@ func proxySetHeader(loc interface{}) string {
-		uniqueCodes = append(uniqueCodes, key)
+	errorLocations := []errorLocation{}
+
+	for upstream := range codesMap {
```
why not 

```
for upstream, dedupedCodes := range codesMap {
```

and then use `dedupedCodes` in below loop instead of `codesMap[upstream]`?

jasongwartz replied:
good point, will do

ElvinEfendi commented on internal/ingress/controller/template/template.go:
```diff
@@ -882,41 +884,87 @@ func proxySetHeader(loc interface{}) string {
+
+	for upstream := range codesMap {
+		codesForUpstream := []int{}
+		for code := range codesMap[upstream] {
```
Could you rewrite it as 

```
for code, _ := range codesMap[upstream] {
```

otherwise it's really confusing to read

jasongwartz replied:
yup, will do

jasongwartz replied:
have done so now, but gofmt throws a warning 'should omit 2nd value'

ElvinEfendi commented on internal/ingress/controller/template/template.go:
```diff
@@ -882,41 +884,87 @@ func proxySetHeader(loc interface{}) string {
-// and returns a unique set (for the template to create @custom_xxx locations)
-func collectCustomErrorsPerServer(input interface{}) []int {
+// and returns a set which is unique by default-upstream and error code. The resulting
+// data structure is a map of upstreamName => []errorCodes, so that sufficiently unique
```
If I read the code correctly this is an intermediary data structure you're using not what the function returns.

ElvinEfendi commented on internal/ingress/controller/template/template.go:
```diff
@@ -882,41 +884,87 @@ func proxySetHeader(loc interface{}) string {
+		sort.Ints(codesForUpstream)
+		errorLocations = append(errorLocations, errorLocation{
+			Name:  upstream,
+			Codes: codesForUpstream,
```
Shoulf we sort the codes as well? Since you're using range at https://github.com/kubernetes/ingress-nginx/pull/3730/files#diff-652deabba49a2a87d4c5a79700dfa1e3R953 when collecting codes, I think their order can change and introduce unnecessary Nginx reloads.

jasongwartz replied:
Unless I made a mistake, that's what I was doing with `sort.Ints()` here: https://github.com/kubernetes/ingress-nginx/pull/3730/files#diff-652deabba49a2a87d4c5a79700dfa1e3R956

ElvinEfendi replied:
My bad, yes you're already doing it 👍

aledbf commented:
@jasongwartz is the PR ready? (I see additional commits since I read it)

ElvinEfendi commented on internal/ingress/controller/controller.go:
```diff
@@ -579,24 +579,21 @@ func (n *NGINXController) getBackendServers(ingresses []*ingress.Ingress) ([]*in
+					sp := location.DefaultBackend.Spec.Ports[0]
+					endps := getEndpoints(location.DefaultBackend, &sp, apiv1.ProtocolTCP, n.store.GetServiceEndpoints)
+					if len(endps) > 0 {
+						klog.V(3).Infof("Using custom default backend for location %q in server %q (Service \"%v/%v\")",
```
As commented earlier, this is not always going to be true. The message should be around:

```
Creating "custom-default-backend-%v" upstream based on default backend annotation
```

jasongwartz replied:
I've moved the logs here around a bit, let me know if there are more improvements we can make

ElvinEfendi commented on internal/ingress/controller/controller.go:
```diff
@@ -579,24 +579,21 @@ func (n *NGINXController) getBackendServers(ingresses []*ingress.Ingress) ([]*in
-								aUpstreams = append(aUpstreams, nb)
-								location.Backend = name
-							}
+				if shouldCreateUpstreamForLocationDefaultBackend(upstream, location) {
```
👍

ElvinEfendi commented on internal/ingress/controller/controller.go:
```diff
@@ -579,24 +579,21 @@ func (n *NGINXController) getBackendServers(ingresses []*ingress.Ingress) ([]*in
+						nb.Endpoints = endps
+						aUpstreams = append(aUpstreams, nb)
+						if len(upstream.Endpoints) == 0 {
+							klog.V(3).Infof("Upstream %q has no active Endpoint", upstream.Name)
```
Log message at https://github.com/kubernetes/ingress-nginx/pull/3730/files#diff-86ac9ff9d75a0c5005c116e766a6127dR586 is more relevant here. Maybe something around mix of the two is even better.

ElvinEfendi commented on internal/ingress/controller/controller.go:
```diff
@@ -579,24 +579,21 @@ func (n *NGINXController) getBackendServers(ingresses []*ingress.Ingress) ([]*in
+						nb.Name = name
+						nb.Endpoints = endps
+						aUpstreams = append(aUpstreams, nb)
+						if len(upstream.Endpoints) == 0 {
```
nitpick: I'd also separate this with a newline from the above logic where we create the upstream for given default backend, and the below if needed we use it as an upstream for this location.

jasongwartz replied:
done

ElvinEfendi commented on internal/ingress/controller/template/template.go:
```diff
@@ -882,41 +884,87 @@ func proxySetHeader(loc interface{}) string {
+}
+
+type errorLocation struct {
+	Name  string
```
nitpick: `UpstreamName` would be more descriptive

jasongwartz replied:
done

ElvinEfendi commented:
I think the CI failure is legit: the test added in this PR expects setting of `default-backend` annotation to cause controller to generate a new Nginx configuration. But that does not happen (we did not need before this PR) because we don't compare `DefaultBackend` attribute when we compare two locations in `types_equals.go`.

So this PR as is requires controller to generate a new Nginx configuration and reload it when `default-backend` annotation changes.

jasongwartz commented:
@ElvinEfendi yup my bad, I knew the CI failure was real and am still working on it - thanks for pointing me in the right direction.
I haven't had so much time recently to look into fixing this up but I will do soon

aledbf commented:
@jasongwartz friendly ping

jasongwartz commented:
@aledbf thanks for the reminder! I addressed the comments and am still working on the e2e tests (I haven't had a lot of time to dive into this but I'll keep at it)

ElvinEfendi commented:
@jasongwartz just comparing `DefaultBackend` when diffing two locations should fix the test issue.

ElvinEfendi commented:
I believe following patch should fix the issue:

```
diff --git a/internal/ingress/types_equals.go b/internal/ingress/types_equals.go
index ff6361faa..0dd6aafdd 100644
--- a/internal/ingress/types_equals.go
+++ b/internal/ingress/types_equals.go
@@ -479,6 +479,10 @@ func (l1 *Location) Equal(l2 *Location) bool {
                return false
        }

+       if l1.DefaultBackend.GetName() != l2.DefaultBackend.GetName() {
+               return false
+       }
+
        return true
 }
```

The reason why I'm not comparing the whole `DefaultBackend` is to avoid unnecessary Nginx reloads when something random other than name in DefaultBackend service changes - because we need a new Nginx configuration only when name changes.

jasongwartz commented:
@ElvinEfendi I should be able to use the new `.ocation.DefaultBackendUpstreamName` for this as well, right? Otherwise I have to do nil-checks before doing `location.DefaultBackend.GetName()`?
I've been testing it locally, and the e2e test still fails, but I'll keep digging

jasongwartz commented:
I'm thinking the failure of the e2e test is because the service of the custom-default-backend doesn't really exist and therefore has no endpoints (so the upstream doesn't get created). I'll try adding a deployment + service and see if that works

ElvinEfendi commented on internal/ingress/types_equals.go:
```diff
@@ -479,6 +479,10 @@ func (l1 *Location) Equal(l2 *Location) bool {
 
+	if l1.DefaultBackendUpstreamName != l2.DefaultBackendUpstreamName {
+		return false
+	}
```
Was this the only change that fixed e2e test failure?

jasongwartz replied:
This, and adding an actual deployment so that endpoints will exist (here: https://github.com/kubernetes/ingress-nginx/pull/3730/files#diff-e82a2942a511f5a12803acc64b3fc676R107)

ElvinEfendi commented on test/e2e/annotations/customhttperrors.go:
```diff
@@ -22,12 +22,17 @@ import (
 	"k8s.io/ingress-nginx/test/e2e/framework"
 )
 
+const defaultBackendName = "upstream-default-backend"
```
this is not being used

jasongwartz replied:
I'll put up a quick follow-up to remove this, wasn't quick enough before merging. thanks for catching it

ElvinEfendi commented:
/lgtm

I'll defer `approve` to @aledbf 

Thank you @jasongwartz !

ElvinEfendi commented:
/hold

oops the PR already has `approved` label

aledbf commented:
/hold cancel
